### ISSUE TITLE
[LINQ] Determine how to implement OfType<T>

### ISSUE THREAD

**Issue by xposure** [beefytech/Beef#834]
.Net has this method `public static IEnumerable<TResult> OfType<TResult>(this IEnumerable source);` which I'm not sure if it will be doable. There is no "clean" way to just have the below code take a single generic and would require you to fill in all of them. I wonder if we could have something like OfType<this TCollection, this TSource, TOf>(..)`. I have run in to this with a couple of them, you have any ideas?
 ```
		public static OfTypeEnumerator<TSource, decltype(default(TCollection).GetEnumerator()), TOf>
			OfType<TCollection, TSource, TOf>(this TCollection items)
			where TCollection : concrete, IEnumerable<TSource>
		{
			return .(items.GetEnumerator());
		}
```

**Comment by bfiete** [comment]
This is a general generic issue also being discussed over here https://github.com/dotnet/csharplang/issues/1349
It's been open for 3 years with no real movement, though.

**Comment by bfiete** [comment]
None of those proposals are THAT great in this context, though. `list.OfType<TOf: String>()`, `list.OfType<,,String>` ... not really great.

**Comment by bfiete** [comment]
Hm, not sure if this is crazy or not- have an `implicit` specifier that means those params will not be expected in the explicit generic arg list.
```
public static OfTypeEnumerator<TSource, decltype(default(TCollection).GetEnumerator()), TOf>
   		OfType<TCollection, TSource, TOf>(this TCollection items)
   		where TCollection : implicit, concrete, IEnumerable<TSource>
   		where TSource : implicit
   	{
   		return .(items.GetEnumerator());
   	}
```

And maybe if there were combined with a feature with named generic arguments then you _could_ specify them by name...

**Comment by xposure** [comment]
I had that same exact idea but doing it on the generic instead of the constraint.

<implicit TSource, ....>

**Comment by xposure** [comment]
Forgot to mention, I don't think its a bad idea either way.

**Comment by bfiete** [comment]
Yeah- your way makes more sense, I think. I'm not 100% sold yet but it does solve a certain class of problems...

**Comment by bfiete** [comment]
Took just over a year, but we do have this now!

```C#
public static TOut Conv<TOut, TIn>(TIn val) where TOut : operator explicit TIn
{
	return (TOut)val;
}

Conv<int,?>(12.34f); // Specify first generic arg and and infer second
Conv<int...>(12.34f); // Specify first generic arg, infer any number of additional generic args after that
```
